router/core/operation_processor.go

```diff
@@ -363,30 +448,182 @@ func (o *OperationKit) Normalize() error {
 	// Set the normalized representation
 	o.parsedOperation.NormalizedRepresentation = o.kit.normalizedOperation.String()
 
-	// Here we copy the normalized variables. After normalization, the variables can be consumed or modified.
-	variablesCopy := make([]byte, len(o.kit.doc.Input.Variables))
-	copy(variablesCopy, o.kit.doc.Input.Variables)
+	return nil
+}
+
+type normalizedOperationCacheEntry struct {
+	operationID              uint64
+	normalizedRepresentation string
+	operationType            string
+	exportedVariables        []byte
+}
 
-	o.parsedOperation.Variables = variablesCopy
+func (o *OperationKit) loadPersistedOperationFromCache() (ok bool, err error) {
 
-	return nil
+	if o.cache == nil {
+		return false, nil
+	}
+
+	cacheKey, ok := o.loadPersistedOperationCacheKey(o.parsedOperation.GraphQLRequestExtensions.PersistedQuery.Sha256Hash)
+	if !ok {
+		return false, nil
+	}
+
+	o.cache.persistedOperationCacheLock.RLock()
+	entry, ok := o.cache.persistedOperationCache[cacheKey]
+	o.cache.persistedOperationCacheLock.RUnlock()
+	if !ok {
+		return false, nil
+	}
+	o.parsedOperation.PersistedOperationCacheHit = true
+	o.parsedOperation.ID = entry.operationID
+	o.parsedOperation.NormalizedRepresentation = entry.normalizedRepresentation
+	o.parsedOperation.Type = entry.operationType
+	if o.parsedOperation.Request.Variables == nil || bytes.Equal(o.parsedOperation.Request.Variables, literalNull) {
+		o.parsedOperation.Request.Variables = []byte("{}")
+	}
+
+	err = jsonparser.ObjectEach(entry.exportedVariables, func(key []byte, value []byte, dataType jsonparser.ValueType, offset int) (err error) {
+		_, exists := o.parsedOperation.VariablesMap[string(key)]
+		if exists {
+			return nil
+		}
+		if dataType == jsonparser.String {
+			stringValue := make([]byte, len(value)+2)
+			stringValue[0] = '"'
+			copy(stringValue[1:], value)
+			stringValue[len(stringValue)-1] = '"'
+			o.parsedOperation.Request.Variables, err = jsonparser.Set(o.parsedOperation.Request.Variables, stringValue, unsafebytes.BytesToString(key))
+		} else {
+			o.parsedOperation.Request.Variables, err = jsonparser.Set(o.parsedOperation.Request.Variables, value, unsafebytes.BytesToString(key))
+		}
+		return
+	})
+	if err != nil {
+		return false, errors.WithStack(fmt.Errorf("failed to set default values for variables: %w", err))
+	}
+
+	o.kit.doc.Input.ResetInputString(entry.normalizedRepresentation)
+	report := &operationreport.Report{}
+	o.kit.parser.Parse(o.kit.doc, report)
+	if report.HasErrors() {
+		return false, &reportError{
+			report: report,
+		}
+	}
+
+	return true, nil
+}
+
+func (o *OperationKit) savePersistedOperationToCache(skipIncludeVariableNames []string, exportedVariables []byte) {
+	if o.cache == nil {
```

Comment:
I'd write it in way that a function call is not made when the cache is not set because the function has no meaning without the cache.